### PR TITLE
Remove ability to set the value of meta fields inside `_source`

### LINKED ISSUE #11051: Enforce meta fields are set through the api, not inside a document
Some meta fields are currently allowed to be set within a document's source:
* `_all`
* `_id`
* `_parent`
* `_routing`
* `_timestamp`
* `_ttl`

In #6730 the ability to set a custom path for `_id` and `_routing` was removed. We should go one step further and not allow setting any meta fields within a document. These are special fields that the system uses for special purposes. There should be one way to set them, through the api.

### PR DESCRIPTION
**Pull request by rjernst**
A few meta fields can currently be set within a document's source.
However, the recommended way to set meta fields like this is through
the api, and setting within the document can be a performance trap
(e.g. needing to find _id in order to route the document).

This change removes the ability to set meta fields within
a document source for 2.0+ indexes.

closes #11051

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -267,8 +267,8 @@ private DocumentMapper merge(DocumentMapper mapper) {
                 List<ObjectMapper> newObjectMappers = new ArrayList<>();
                 List<FieldMapper<?>> newFieldMappers = new ArrayList<>();
                 for (RootMapper rootMapper : mapper.mapping().rootMappers) {
-                    if (!rootMapper.includeInObject() && rootMapper instanceof FieldMapper) {
-                        newFieldMappers.add((FieldMapper<?>) rootMapper);
+                    if (rootMapper instanceof FieldMapper) {
+                        newFieldMappers.add((FieldMapper)rootMapper);
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/MapperService.java]: s/FieldMapper/FieldMapper<?>/ ?

File: src/main/java/org/elasticsearch/index/mapper/Mapping.java [hunk 2]
```diff
@@ -38,6 +42,8 @@
  */
 public final class Mapping implements ToXContent {
 
+    public static final List<String> legacyIncludeInObject = Arrays.asList("_all", "_id", "_parent", "_routing", "_timestamp", "_ttl");
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/index/mapper/Mapping.java]: I think we generally use capitals for constants?

**Comment by jpountz** [comment]
This looks great! I think we should also add a note about this to the migration guide and have a simple test that checks that specifying these fields still works on old indices?

**Comment by rjernst** [comment]
@jpountz I pushed a commit adding a migration note and tests. I actually found while testing that only `_id` and `_parent` actually worked when specified in a document.  The others would result in ignoring the field, which I am testing for true "backcompat".

**Comment by jpountz** [comment]
LGTM

**Pull request merged**